crates/bevy_ecs/src/query/access.rs

```diff
@@ -428,23 +378,22 @@ impl<T: SparseSetIndex> Access<T> {
             other.component_read_and_writes_inverted,
         ) {
             (true, true) => {
-                self.component_read_and_writes
-                    .intersect_with(&other.component_read_and_writes);
+                self.component_read_and_writes = self
+                    .component_read_and_writes
+                    .intersection(&other.component_read_and_writes)
```

Comment:
If it helps, I think you can use `retain` to take care of shuffling the elements around, like:

```rust
fn intersect_with(&mut self, other: &Self) {
    let mut j = 0;
    self.0.retain(|current| {
        // Advance past any smaller elements in other
        while j < other.len() && other.0[j] < *current { j += 1; }
        // It's only in the intersection if it's in other,
        // and this is the only place in other it could be
        j < other.len() && other.0[j] == *current
    });
}
```

`difference_with` would be the same but with a `!` in front of the condition.

Reply:
Very clever, thank you